client/state/terms/reducer.js

```diff
@@ -144,6 +147,28 @@ export function items( state = {}, action ) {
 				}
 			} );
 
+		case TERMS_ADD_REQUEST:
+			return merge( {}, state, {
+				[ action.siteId ]: {
+					[ action.taxonomy ]: {
+						[ action.temporaryId ]: action.term
+					}
+				}
+			} );
+
+		case TERMS_ADD_REQUEST_SUCCESS:
+			const taxonomyTerms = merge( {}, state );
+			taxonomyTerms[ action.siteId ][ action.taxonomy ][ action.term.ID ] = action.term;
+			delete taxonomyTerms[ action.siteId ][ action.taxonomy ][ action.temporaryId ];
+
+			return taxonomyTerms;
+
+		case TERMS_ADD_REQUEST_FAILURE:
+			const terms = merge( {}, state );
```

Comment:
Having issues with the naming conflicts of `taxonomyTerms` above? Two solutions to consider:

- [`createReducer`](https://github.com/Automattic/wp-calypso/blob/58ea5b0a78430cbd4e7b4ccf5d78dee06f41adad/client/state/utils.js#L27-L56) has a side-effect in that each action type's function has its own scope
- I've been using a standard [JavaScript block](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/block), which is uncommon only because it was previously ineffective for `var` declarations, but works great with `const` / `let` ([example usage](https://github.com/Automattic/wp-calypso/blob/58ea5b0a78430cbd4e7b4ccf5d78dee06f41adad/client/state/posts/reducer.js#L142-L160))